### PR TITLE
Refactoring rocmlir-gen.

### PR DESCRIPTION
**Pull request by pcf000**
Originally for #750, but #750 changes can be independent of refactoring.

I started abstracting functions and cleaning up some old things, to help me figure out what was going on.  Then the changes for #750 turned out to be simpler than I expected, and can work without the refactoring.  If this is still useful, we can keep it.

### CODE REVIEW

File: mlir/test/rocmlir-driver/conv2d_harness/conv2d_harness.mlir [hunk 1]
```diff
@@ -1,6 +1,6 @@
-// RUN: rocmlir-gen --arch %arch -p --host %s | FileCheck %s --check-prefix=HARNESS
-// RUN: rocmlir-gen --arch %arch -p --host %s | rocmlir-driver -c | FileCheck %s --check-prefix=LOWERING
-// RUN: rocmlir-gen --arch %arch -p --host %s | rocmlir-driver -c | mlir-cpu-runner -O2 --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext --entry-point-result=void | FileCheck %s --check-prefix=E2E
+// RUN: rocmlir-gen --arch %arch -p %s | FileCheck %s --check-prefix=HARNESS
```
Comment (pcf000) [inline mlir/test/rocmlir-driver/conv2d_harness/conv2d_harness.mlir:1]: Turns out that --host didn't do anything.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 2]
```diff
@@ -388,11 +388,6 @@ static llvm::cl::opt<bool> emitTuningKey(
 //////////////////////////////////////////////////////////////////////////
 
 // generate host harness program.
-static llvm::cl::opt<bool>
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:391]: --host, aka readHostHarness, never did anything.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 3]
```diff
@@ -2469,21 +2464,10 @@ static LogicalResult populateHostHarnessLogic(
   // Construct main function.
   auto func = func::FuncOp::create(loc, "main", b.getFunctionType({}, {}));
   module.push_back(func);
-
-  // Construct a new Block.
   Block *block = func.addEntryBlock();
   b.setInsertionPoint(block, block->begin());
 
-  llvm::SmallDenseMap<int32_t, Value> i32vals;
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:2477]: Moved closer to use.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 4]
```diff
@@ -2686,218 +2679,162 @@ static LogicalResult populateHostHarnessLogic(
...
-                      arith::ArithDialect, vector::VectorDialect,
-                      gpu::GPUDialect, linalg::LinalgDialect>();
-
-  // Parse pass names in main to ensure static initialization completed.
-  llvm::cl::ParseCommandLineOptions(argc, argv,
-                                    "MLIR Rock Dialect host generation\n");
-
-  OpBuilder builder(&context);
-  ModuleOp module;
+static ModuleOp readTestFile(std::string inputFilenameStr, bool &hasUserKernel,
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:2673]: Abstracted the file-reading logic into its own function.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 5]
```diff
@@ -2686,218 +2679,162 @@ static LogicalResult populateHostHarnessLogic(
...
+  return module;
+}
 
-    module.walk([&](func::FuncOp func) -> WalkResult {
-      if (func->hasAttr("kernel")) {
-        hasUserKernel = true;
-      }
-      return WalkResult::advance();
-    });
+static ModuleOp generateKernel(MLIRContext *context, GenParams &genParams,
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:2718]: Abstracted the kernel-generation portion of main().

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 6]
```diff
@@ -2927,6 +2864,51 @@ int main(int argc, char **argv) {
...
+  }
+  populateDefaults();
+
+  bool hasUserKernel = !testFuncName.empty();
+
+  ModuleOp module;
+  GenParams genParams;
+
+  if (!inputFilename.empty()) {
+    module = readTestFile(inputFilename.getValue(), hasUserKernel, &context);
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:2888]: Call the file-reading function here.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 7]
```diff
@@ -2927,6 +2864,51 @@ int main(int argc, char **argv) {
...
+      llvm::errs()
+          << "Clone validation is not compatible with kernel generation.\n";
+      exit(1);
+    }
+    OpBuilder builder(&context);
+    module = ModuleOp::create(builder.getUnknownLoc());
+  }
+
+  if (!hasUserKernel) {
+    module = generateKernel(&context, genParams, module);
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:2900]: Call the kernel-generation function here.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 8]
```diff
@@ -2507,6 +2491,15 @@ static LogicalResult populateHostHarnessLogic(
   bool isRandom = (randomSeed != "fixed" && randomSeed != "none");
 
   if (isRandom) {
+    llvm::SmallDenseMap<int32_t, Value> i32vals;
+    auto getI32Val = [&](int32_t v) {
```
Comment (krzysz00) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: This whole thing could probably be replaced by `createOrFold`

**Review by krzysz00 — APPROVED** [review]
These all look like reasonable cleanups

**Pull request merged**